hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsRestOperation.java

```diff
@@ -347,7 +347,7 @@ private boolean executeHttpOperation(final int retryCount,
 
     LOG.debug("HttpRequest: {}: {}", operationType, httpOperation);
 
-    if (client.getRetryPolicy().shouldRetry(retryCount, httpOperation.getStatusCode())) {
+    if (client.getRetryPolicy(failureReason).shouldRetry(retryCount, httpOperation.getStatusCode())) {
```

Comment:
Can we avoid having client.getRetryPolicy(failureReason). Can there be a way where each iteration knows which retryPolicy obj to be used. Instance of problem here:
1. iteration0 failed with connectionTimeout, at line 350, failureReason would be CT and it would fire for LinearRetryPolicy.
2. But at this point, it has got lets say 503. Som 352 become lets say 503_ING.
3. Now the sleep would happen as per exponentialRetryPolicy.
4. Problem here being, shouldRetry fired as per LinearRetryPolicy, and sleep happened as per ExponentialRetryPolicy.

Reply:
+1 on this

Reply:
If iteration0 fails then line number 350 should not hit. Failure will be caught by one of the exceptions and inside catch failureReason will be CT. Inside catch itself shouldRetry() will be called for LinearRetry and it will return false here itself.

I think I am missing something here. Please correct me if I am wrong.

Reply:
say you are on retry=1; retry=0 failed becasue of CT. at line 350, you have failureReason=CT, though you are in retry=1. client.getRetryPolicy() will give as per CT and not by what has happened on retry=1. But when you go on the loop and for the thread.sleep, you would sleep as per failure on retry=1.